### ISSUE TITLE
New default behavior for rules in 5.0?

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#10215]
Wondering if there are rules whose default behavior we should consider changing in ESLint 5.0. I'm mostly thinking of rules that had a new option introduced in a semver-minor release, but which we think should be the new default behavior in a semver-major release.

Anyone have any thoughts/ideas?

**Comment by ljharb** [comment]
- https://eslint.org/docs/rules/comma-dangle, enable "functions"?
 - https://eslint.org/docs/rules/padded-blocks, make always/never include switch statements?
 - https://eslint.org/docs/rules/object-curly-newline change the default from "multiline" to "consistent"?
 - https://eslint.org/docs/rules/max-len enable all the ignore options by default?

**Comment by Aladdin-ADD** [comment]
- https://eslint.org/docs/rules/no-empty "allowEmptyCatch": true?
- https://eslint.org/docs/rules/no-self-assign "props": true?

**Comment by not-an-aardvark** [comment]
> - https://eslint.org/docs/rules/object-curly-newline change the default from "multiline" to "consistent"?

:+1:

> > - https://eslint.org/docs/rules/max-len enable all the ignore options by default?
>
> `ignoreRegExpLiterals: true` looks reasonable since Regular expression literals disallow line breaks. I'm not sure about other `ignore***` options.

I think the default behavior of `max-len` should be to enforce a maximum line length, without any caveats. Changing this seems like it would violate the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). (It's not always the case that users want to ignore RegExp literals -- for some projects the solution to a long regex would either be "simplify it to make it shorter" or "use the `RegExp` constructor on concatenated string literals.)

> - https://eslint.org/docs/rules/no-self-assign "props": true?

:+1:

> - https://eslint.org/docs/rules/no-undef "typeof": true? (refs #10203 )

I'm not sure about this. It's pretty common for code to do something like:

```js
if (typeof possiblyExistingGlobal !== 'undefined') {
  doSomethingWith(possiblyExistingGlobal);
}
```

In this case it doesn't make sense to specify `possiblyExistingGlobal` as a global in configuration, because the user would still want to disallow it outside of blocks that are guarded with a check for its existence. It seems like it would be annoying to add an `eslint-disable` comment for every `if` statement like this.* Given that most ESLint users use `no-undef`, I'm leaning towards not wanting to change this default because it would cause a lot of ecosystem churn from false positives.

*Currently, it's necessary to use an `eslint-disable` comment anyway for every use of `possiblyExistingGlobal` inside the block. I think we should fix this (there have been a few proposals about it in the past), but it's a separate issue.

**Comment by ljharb** [comment]
> use the RegExp constructor on concatenated string literals.

that would be both astonishing and a bad practice.

**Comment by not-an-aardvark** [comment]
As evidence that this is actually a thing that people do, here are some examples from the `nodejs/node` codebase, which used to use `max-len` without `ignoreRegExpLiterals` (although the config was changed later):

[1](https://github.com/nodejs/node/blob/65d97c96aa59517f90c34bf5a6b962e1b5e42316/test/parallel/test-buffer-alloc.js#L947-L948), [2](https://github.com/nodejs/node/blob/65d97c96aa59517f90c34bf5a6b962e1b5e42316/test/parallel/test-buffer-writeint.js#L10-L11), [3](https://github.com/nodejs/node/blob/65d97c96aa59517f90c34bf5a6b962e1b5e42316/test/pummel/test-crypto-dh.js#L42-L43), [4](https://github.com/nodejs/node/blob/65d97c96aa59517f90c34bf5a6b962e1b5e42316/test/pummel/test-crypto-dh.js#L51-L52)

Opinions clearly vary about whether this style is appealing. Still, I think configuring something like `max-len: [error, 80]`, the user is telling ESLint that they would like their line length to be capped at 80. Maybe this isn't actually what the user wants, but I don't think ESLint should silently override what the user said and enforce something different. If it turns out that the user wanted to ignore regex literals, they'll find out soon enough when an error is reported, and they can change their config.

In my view, the user expectation in this case is different from the expectation when ESLint chooses defaults for most other rules, because the idea of enforcing a maximum line length is very simple. The user probably has a very clear idea of what they're trying to enforce when configuring `max-len`, in contrast to other rules where the user probably hasn't considered all the cases that the rule needs to deal with.

**Comment by not-an-aardvark** [comment]
**TSC Summary**: This issue proposes changing the default behavior of some rules in ESLint v5.0. 

**TSC Question**: Which rule defaults should we change?

**Comment by not-an-aardvark** [comment]
In [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/adef4bed53b4e3160c19e82e7736ee651f540ee0/notes/2018/2018-05-10.md#new-default-behavior-for-rules-in-50), the TSC decided to change the default for `object-curly-newline` to `"consistent"`, and to change the default for `no-self-assign` to `"props": true`.
